protocols/raft/src/main/java/io/atomix/protocols/raft/roles/FollowerRole.java

```diff
@@ -202,37 +172,80 @@ private void sendPollRequests() {
   @Override
   public CompletableFuture<InstallResponse> onInstall(InstallRequest request) {
     CompletableFuture<InstallResponse> future = super.onInstall(request);
-    resetHeartbeatTimeout();
+    resetHeartbeatTimeoutFromDifferentThread();
     return future;
   }
 
   @Override
   public CompletableFuture<ConfigureResponse> onConfigure(ConfigureRequest request) {
     CompletableFuture<ConfigureResponse> future = super.onConfigure(request);
-    resetHeartbeatTimeout();
+    resetHeartbeatTimeoutFromDifferentThread();
     return future;
   }
 
   @Override
   public CompletableFuture<AppendResponse> onAppend(AppendRequest request) {
+    // Reset the heartbeat timeout.
+    resetHeartbeatTimeoutFromDifferentThread();
+
+    return super.onAppend(request);
+  }
+
+  @Override
+  public void onLeaderHeartbeat(LeaderHeartbeatRequest request) {
+    logRequest(request);
+    // Reset the heartbeat timeout.
+    resetHeartbeatTimeoutFromHeartbeatThread();
+  }
+
+  private void resetHeartbeatTimeoutFromDifferentThread() {
+    raft.runOnHeartbeatContext(this::resetHeartbeatTimeoutFromHeartbeatThread);
+  }
+
+  /**
+   * Resets the heartbeat timer.
+   */
+  private void resetHeartbeatTimeoutFromHeartbeatThread() {
+    raft.checkHeartbeatThread();
+    if (!isRunning()) {
```

Comment:
I think `isRunning` is checking a non-volatile bool which is probably set from the Raft thread, so it might be that we aren't anymore but this still returns true?

Reply:
Good catch